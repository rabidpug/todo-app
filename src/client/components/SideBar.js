import {
  Icon,
  Layout,
  Menu,
} from 'antd';

import React from 'react';

const { Item, SubMenu, } = Menu;
const { Sider, } = Layout;
const SideBar = ( {
  isSidebarCollapsed, menuItems, goToPath, toggleSideBar, currentPath, toggleKey, openKeys,
} ) => {
  const selectedKeys = [];
  const openSelectedKeys = [ ...openKeys, ];
  const determineIfOpen = item => {
    if ( currentPath.indexOf( item.path ) === 0 && item.path.length >= currentPath.length - 1 ) {
      if ( item.subMenu ) {
        openSelectedKeys.push( item.key );

        item.subMenu.forEach( determineIfOpen );
      } else selectedKeys.push( item.key );
    }
  };

  menuItems.forEach( determineIfOpen );

  const menuItemMap = item =>
    item.subMenu ? (
      <SubMenu
        key={ item.key }
        onTitleClick={ ( { key, } ) => toggleKey( key ) }
        title={
          <span>
            <Icon type={ item.icon } />
            <span>{item.label}</span>
          </span>
        }>
        {item.subMenu.map( menuItemMap )}
      </SubMenu>
    ) : (
      <Item
        key={ item.key }
        path={ item.path }>
        <Icon type={ item.icon } />
        <span>{item.label}</span>
      </Item>
    );

  return (
    <Sider
      collapsed={ isSidebarCollapsed }
      collapsible
      onMouseEnter={ () => isSidebarCollapsed && toggleSideBar() }
      onMouseLeave={ () => !isSidebarCollapsed && toggleSideBar() }
      style={ { backgroundColor: '#fff', } }
      trigger={ null }>
      <div className='logo' />
      <Menu
        mode='inline'
        onClick={ ( { item, } ) => goToPath( item.props.path ) }
        openKeys={ isSidebarCollapsed ? [] : openSelectedKeys }
        selectedKeys={ selectedKeys }
        style={ { height: '100%', } }
        theme='dark'>
        {menuItems.map( menuItemMap )}
      </Menu>
    </Sider>
  );
};

export default SideBar;
