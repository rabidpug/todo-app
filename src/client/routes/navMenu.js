import HomeSideBar from 'Containers/HomeSideBar';

const navMenu = [
  {
    component : HomeSideBar,
    menuItems : [
      {
        icon    : 'home',
        key     : 'sub1',
        label   : 'Home',
        path    : '/',
        subMenu : [
          {
            icon  : 'home',
            key   : 'subitem1',
            label : 'Home Page',
            path  : '/',
          },
        ],
      },
      {
        icon  : 'smile-o',
        key   : 'item1',
        label : 'Welcome',
        path  : '/welcome',
      },
      {
        icon  : 'frown-o',
        key   : 'item2',
        label : 'Bye',
        path  : '/bye',
      },
      {
        icon  : 'file-text',
        key   : 'item3',
        label : 'Todo',
        path  : '/todo',
      },
    ],
    path: '*',
  },
];

export default navMenu;
