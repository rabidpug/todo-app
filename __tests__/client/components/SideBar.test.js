import React from 'react';
import SideBar from 'Components/SideBar';
import navMenu from 'Routes/navMenu';
import { shallow, } from 'enzyme';
import { shallowToJson, } from 'enzyme-to-json';

describe(
  'SideBar', () => {
    it(
      'Should render the SideBar', () => {
        const [ { menuItems, }, ] = navMenu;
        const wrapper = shallow( <SideBar
          currentPath='/'
          menuItems={ menuItems }
          openKeys={ [] } /> );

        expect( shallowToJson( wrapper ) ).toMatchSnapshot();
      }
    );
  }
);
