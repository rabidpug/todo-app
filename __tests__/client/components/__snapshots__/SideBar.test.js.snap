// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar Should render the SideBar 1`] = `
<Sider
  collapsedWidth={80}
  collapsible={true}
  defaultCollapsed={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  prefixCls="ant-layout-sider"
  reverseArrow={false}
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
  trigger={null}
  width={200}
>
  <div
    className="logo"
  />
  <Menu
    className=""
    mode="inline"
    onClick={[Function]}
    openKeys={
      Array [
        "sub1",
      ]
    }
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        "subitem1",
      ]
    }
    style={
      Object {
        "height": "100%",
      }
    }
    theme="dark"
  >
    <SubMenu
      key="sub1"
      onTitleClick={[Function]}
      title={
        <span>
          <Icon
            type="home"
          />
          <span>
            Home
          </span>
        </span>
      }
    >
      <MenuItem
        key="subitem1"
        path="/"
      >
        <Icon
          type="home"
        />
        <span>
          Home Page
        </span>
      </MenuItem>
    </SubMenu>
    <MenuItem
      key="item1"
      path="/welcome"
    >
      <Icon
        type="smile-o"
      />
      <span>
        Welcome
      </span>
    </MenuItem>
    <MenuItem
      key="item2"
      path="/bye"
    >
      <Icon
        type="frown-o"
      />
      <span>
        Bye
      </span>
    </MenuItem>
    <MenuItem
      key="item3"
      path="/todo"
    >
      <Icon
        type="file-text"
      />
      <span>
        Todo
      </span>
    </MenuItem>
  </Menu>
</Sider>
`;
